### ISSUE TITLE
Thread and Fiber whitelisting

### ISSUE THREAD

**Issue by codegourmet** [ruby-prof/ruby-prof#188]
In issue #178 there was some discussion on threaded profiling.
One solution was to set `::RubyProf.exclude_threads` to all threads except the current one.

This is not viable in a busy webserver scenario, since new threads will be added during the current profile run and will show up because they have not been excluded up front. So there's no way to isolate the originally targeted request.
In this case we need whitelisting, not blacklisting.

I implemented whitelisting one thread (and optionally one fiber), but only for the `initialize()` method and in a rather rudimentary way. It works just fine in the tests and in my real-life puma webserver scenario.

Here's an example from a patched `Rack::RubyProf` middleware, whitelisting the current request via thread and fiber:
```
data = ::RubyProf::Profile.profile(
  RubyProf::WALL_TIME, [], Thread.current, Fiber.current
) do
  result = @app.call(env)
end
```

I did a pull request (#189), but since there's still so much to do here and I'm unsure if this patch isn't considered feature creep feel free to cancel it anytime.

Further recommended TODOs:
- check if the code is OK; I'm new to rubyprof, threading, fibers, C extensions and my C is rusty. So this solution might have some problems I'm not aware of!
- adapt `compatibility.rb` so we have static setters for `whitelist_thread` and `whitelist_fiber`, like `exclude_threads` (for now I only modified the C extension `initialize` method)
- implement `whitelist_threads` and `whitelist_fibers` as arrays, like `exclude_threads`
- empty or `nil` values should deactivate the feature (for now, if you want to whitelist a fiber you'll also have to whitelist a thread)
- documentation

**Comment by skaes** [comment]
I've been working on a similar refactoring, but not pushed yet, all related to #178. I'll a look at your proposal and compare the approaches. Hopefully next weekend.

**Comment by skaes** [comment]
I have completed my implementation. I haven't added whitelisting fibers yet.

I have changed the signature of the Profile initialize to accept an options hash.

Please have a look at the current master and let me know if this code fulfills your need.

**Comment by codegourmet** [comment]
Sorry for the delay, busybusy. I'm testing right now. Will report tomorrow.
The new changes blow my mind, awesome!

BTW: whitelisting fibers is not necessary in the rack case, this was a misunderstanding/lack of knowledge on my side. merging fibers is very useful, I will test it out.

**Comment by codegourmet** [comment]
I deployed the version with the new signature for Profile#initialize and it seems to be looking OK, thanks for your work on this feature!
I used `[Thread.current]` for whitelisting, enabled `merge_fibers` and WALL_TIME.

Use case was profiling middleware for a high-load puma application with multiple workers and threads.

I will cancel my pull request (#189) concerning this.
Should I close this issue?
